app/src/main/java/ai/kun/opentrace/worker/BLETrace.kt

```diff
@@ -0,0 +1,95 @@
+package ai.kun.opentrace.worker
+
+import ai.kun.opentrace.util.Constants
+import ai.kun.opentrace.util.Constants.PREF_UNIQUE_ID
+import ai.kun.opentrace.util.Constants.RANGE_ENVIRONMENTAL
+import android.app.AlarmManager
+import android.bluetooth.BluetoothGattServer
+import android.bluetooth.BluetoothManager
+import android.bluetooth.le.BluetoothLeAdvertiser
+import android.bluetooth.le.BluetoothLeScanner
+import android.content.Context
+import android.content.SharedPreferences
+import java.util.*
+import kotlin.math.pow
+
+
+object BLETrace {
+    private val mBleServer : BLEServer = BLEServer()
+    private val mBleClient : BLEClient = BLEClient()
+
+    private var isInit = false
+    lateinit var context : Context
+    lateinit var bluetoothGattServer: BluetoothGattServer
+    lateinit var bluetoothManager: BluetoothManager
+    lateinit var bluetoothLeScanner: BluetoothLeScanner
+    lateinit var bluetoothLeAdvertiser: BluetoothLeAdvertiser
+    lateinit var alarmManager: AlarmManager
+
+
+    private var uniqueID: String? = null
+    fun id(context: Context): String? {
+        synchronized(this) {
+            if (uniqueID == null) {
+                val sharedPrefs = context.getSharedPreferences(
+                    PREF_UNIQUE_ID, Context.MODE_PRIVATE
+                )
+                uniqueID = sharedPrefs.getString(PREF_UNIQUE_ID, null)
+                if (uniqueID == null) {
+                    uniqueID = UUID.randomUUID().toString()
```

Comment:
If it's for uniquely identifying bluetooth device I think we should use uid coming from Firebase. That way we can assign user to a device easily.

Reply:
I thought long and hard about that one.  The reason I did it this way is that on the back end we can easily associate the two and this system allows us to identify a device even if the user doesn't sign in.  I wanted to capture both, but that was going to complicate the BLE code, so I went with just the device ID.  If product wants to associate the two Firebase will know both.  If product decides that they want to support tracking users that want to remain anonymous, this is the best solution because once you delete the application from your phone the source of the UUID is wiped out.  It allows us to have a bunch of information on the cloud side, but still puts the identity in the hands of the user because the other devices see only what amounts to an installation ID.

Reply:
As we agreed I'll add anonymous login and then change this in my next PR which will add to the login flow.